source/common/upstream/health_discovery_service.cc

```diff
@@ -68,10 +165,73 @@ void HdsDelegate::onReceiveTrailingMetadata(Http::HeaderMapPtr&& metadata) {
 
 void HdsDelegate::onRemoteClose(Grpc::Status::GrpcStatus status, const std::string& message) {
   ENVOY_LOG(warn, "gRPC config stream closed: {}, {}", status, message);
-  response_timer_->disableTimer();
+  server_response_timer_->disableTimer();
   stream_ = nullptr;
   handleFailure();
 }
 
+HdsCluster::HdsCluster(Runtime::Loader& runtime, const envoy::api::v2::Cluster& cluster,
+                       const envoy::api::v2::core::BindConfig& bind_config, Stats::Store& stats,
+                       Ssl::ContextManager& ssl_context_manager,
+                       Secret::SecretManager& secret_manager, bool added_via_api,
+                       ClusterInfoFactory& info_factory)
+    : runtime_(runtime), cluster_(cluster), bind_config_(bind_config), stats_(stats),
+      ssl_context_manager_(ssl_context_manager), secret_manager_(secret_manager),
+      added_via_api_(added_via_api), initial_hosts_(new HostVector()) {
+  ENVOY_LOG(debug, "Creating an HdsCluster");
+  priority_set_.getOrCreateHostSet(0);
+
+  info_ = info_factory.createClusterInfo(runtime_, cluster_, bind_config_, stats_,
+                                         ssl_context_manager_, secret_manager_, added_via_api_);
+
+  for (const auto& host : cluster.hosts()) {
+    initial_hosts_->emplace_back(HostSharedPtr{
+        new HostImpl(info_, "", Network::Address::resolveProtoAddress(host),
+                     envoy::api::v2::core::Metadata::default_instance(), 1,
+                     envoy::api::v2::core::Locality().default_instance(),
+                     envoy::api::v2::endpoint::Endpoint::HealthCheckConfig().default_instance())});
+  }
+  initialize([] {});
+}
+
+ClusterSharedPtr HdsCluster::create() { NOT_IMPLEMENTED_GCOVR_EXCL_LINE; }
+
+HostVectorConstSharedPtr HdsCluster::createHealthyHostList(const HostVector& hosts) {
+  HostVectorSharedPtr healthy_list(new HostVector());
```

Comment:
Would it make sense to just inherit from `ClusterImplBase` to get some of these methods for free?